library/oid.c

```diff
@@ -253,42 +253,96 @@ FN_OID_GET_ATTR1(mbedtls_oid_get_attr_short_name, oid_x520_attr_t, x520_attr, co
 typedef struct {
     mbedtls_oid_descriptor_t    descriptor;
     int                 ext_type;
+    int                 is_supported;
 } oid_x509_ext_t;
 
 static const oid_x509_ext_t oid_x509_ext[] =
 {
     {
-        { ADD_LEN( MBEDTLS_OID_BASIC_CONSTRAINTS ),    "id-ce-basicConstraints",    "Basic Constraints" },
-        MBEDTLS_OID_X509_EXT_BASIC_CONSTRAINTS,
+        { ADD_LEN( MBEDTLS_OID_AUTHORITY_KEY_IDENTIFIER ),    "id-ce-authorityKeyIdentifier",      "Authority Key Identifier" },
+        MBEDTLS_OID_X509_EXT_AUTHORITY_KEY_IDENTIFIER, 0,
```

Comment:
This increases the size of `oid.o` significantly by the standards of small embedded platforms. In a Cortex-M0 build in the default configuration, I get 5892 bytes before, 6673 bytes after.

How about changing the interface so that the extra OIDs are only defined if `MBEDTLS_X509_ALLOW_UNSUPPORTED_CRITICAL_EXTENSION` is enabled, and this allows all unsupported extensions by default (for backward compatibility) but the set of permitted extensions can be changed at runtime?

Reply:
Sorry, I am confused.

In the current model, `MBEDTLS_X509_ALLOW_UNSUPPORTED_CRITICAL_EXTENSION` doesn't require any extra OIDs. Extra OIDs are needed to get extension type from asn1_buf and supported/unsupported flag.

If one wants to permit certain unsupported extension in runtime and obviously doesn't define `MBEDTLS_X509_ALLOW_UNSUPPORTED_CRITICAL_EXTENSION`. how do we recognize extension type without extra OIDs?

Reply:
You can parse the ASN.1 sequence crt->v3_ext. But that's obviously not nice at all. Ideally I'd like to come up with a solution that's reasonably easy to use, doesn't add a new compile-time option (or maybe even better: makes the existing one moot), and doesn't increase the code size for embedded builds. Maybe the library could provide a pre-broken-down sequence for the extensions, in such a way that the application can easily look up the extension with a given OID? Or the application could provide a list of OID to accept even if critical?

Since we now have a mailing list, I've started a design discussion there: https://lists.trustedfirmware.org/pipermail/mbed-tls/2020-April/000009.html . Feel free to reply here or on the mailing list.

Reply:
> Ideally I'd like to come up with a solution that's reasonably easy to use, doesn't add a new compile-time option (or maybe even better: makes the existing one moot), and doesn't increase the code size for embedded builds.

I think I might have come up with such a solution, please see https://github.com/ARMmbed/mbedtls/pull/3243